### PR TITLE
fix(transaction): Fix namespace access

### PR DESCRIPTION
**Pull request by dranikpg**
Our area of attack during concurrent transaction access is the call to `DisarmInShard` and `DisarmInShardWhen`, which only access is_armed - an atomic varible. It is not safe to arbitrarily call GetNamespace() if we write to it in InitBase

Solution: Don't write to it post first initialization

### CODE REVIEW

**Comment by dranikpg** [comment]
This is just the easiest solution that I see, definitely not the most elegant one. We currently don't have any guards that would prevent changing namespaces within a single transaction, but like we see with the blocking flow, allowing this has to be carefully considered

File: src/server/transaction.cc [hunk 1]
```diff
@@ -176,11 +176,17 @@ Transaction::~Transaction() {
...
 void Transaction::InitBase(Namespace* ns, DbIndex dbid, CmdArgList args) {
   global_ = false;
-  namespace_ = ns;
   db_index_ = dbid;
   full_args_ = args;
   local_result_ = OpStatus::OK;
   stats_.coordinator_index = ProactorBase::me() ? ProactorBase::me()->GetPoolIndex() : kInvalidSid;
+
+  // Namespace is read by poll execution, so it can't be changed on the fly
+  if (IsScheduled()) {
```
Comment (romange) [inline src/server/transaction.cc:185]: Is it possible to check if ns_ is the same as ns when ns_ is not null?

**Review by romange — APPROVED** [review]

**Review by chakaz — APPROVED** [review]
Thanks for looking into this!

File: src/server/transaction.cc (hunk 1 above)
Comment (romange) [reply src/server/transaction.cc:185]: i.e. `if (ns_ != nullptr)  { DCHECK(ns_ == ns); }`

Comment (chakaz) [reply src/server/transaction.cc:185]: Maybe `DHECK(ns_ == nullptr || ns_ == ns);`

**Review by chakaz — APPROVED** [review]

**Pull request merged**